File: C++/Divide-and-Conquer/karatsuba.cpp
// =========================== Problem Statement ========================
/*
The problem that the above code is solving is to multiply two 
large integers using the Karatsuba algorithm. The Karatsuba 
algorithm is a fast multiplication algorithm that uses a divide 
and conquer strategy to compute the product of two numbers. It is 
based on the observation that the product of two numbers can be 
computed more efficiently by recursively computing the product of 
their halves and then using these partial products to compute the final result.

Here is the formal problem statement:

Input: Two integers a and b with at most n digits, where n is a power of 2.

Output: The product of a and b.

The Karatsuba algorithm has a time complexity of O(n^1.585), which is 
faster than the traditional O(n^2) time complexity of simple multiplication 
algorithms. This makes it especially useful for multiplying very large integers 
or for applications where performance is critical.

*/

// =========================== CODE with C++ ========================
#include <iostream>
#include <algorithm>
#include <cstring>
#include<math.h>

using namespace std;

// Function to compute a*b using Karatsuba algorithm
long long karatsuba(long long a, long long b)
{
    // Base case: return a*b if a or b is less than 10
    // This is the stopping condition for the recursive calls
    if (a < 10 || b < 10)
        return a*b;

    // Calculate the length of the two numbers
    int n = max(to_string(a).length(), to_string(b).length());

    // Divide n by 2
    int n2 = n / 2;

    // Split the numbers into high and low digits
    // x and z are the high digits, y and w are the low digits
    long long x = a / (long long)pow(10, n2);
    long long y = a % (long long)pow(10, n2);
    long long z = b / (long long)pow(10, n2);
    long long w = b % (long long)pow(10, n2);

    // Recursively compute p, q, and r
    // p = x*z
    // q = (x+y)*(z+w)
    // r = y*w
    long long p = karatsuba(x, z);
    long long q = karatsuba(x + y, z + w);
    long long r = karatsuba(y, w);

    // Return the final result using the Karatsuba formula
    // a*b = (10^(2*n2))*p + (10^n2)*(q-p-r) + r
    return (long long)pow(10, 2 * n2) * p + (long long)pow(10, n2) * (q - p - r) + r;
}

int main()
{
    // Read in the two numbers to be multiplied
    long long a, b;
    cin >> a >> b;

    // Call the karatsuba function and print the result
    cout << karatsuba(a, b) << endl;
    return 0;
}

File: Python/Divide-and-Conquer/Karatsuba.py
"""
------------------------------------------- The Problem Statement ------------------------------------------------------

    INPUT :- Two numbers number1 and number2 with n digits where n is a power of 2

    OUTPUT :- The product of number1 and number2 

----------------------------------------------- Complexities ---------------------------------------------------------------

    
    Time complexity in using asterisk   = BigO(n^2)
    Time Complexity of Karatsuba Alg0   = BigO(n^1.58)

"""
# Function being called or the function executing the karatsuba function
def karatsuba(num1,num2):

  # BASE CASE of Recursion
    if(num1<10 or num2<10):
        return num1*num2

    else:
        num1_string = str(num1) # Changing to string to find the length of the number1
        num2_string = str(num2) # Changing to string to find the length of number2

        length = max(len(num1_string), len(num2_string)) # Finding max length
        half=int(length/2)  # Splitting the number into two halves we need the length to be divided by 2

        # Finding a and c i.e. the higher bits for each number
        a = int(num1_string[:-half])
        c = int(num2_string[:-half])

        # Finding b and d i.e. the lower bits for each number
        b = int(num1_string[-half:])
        d = int(num2_string[-half:])

        # Finding ac, bd and ad_plus_bc
        ac = karatsuba(a, c)
        bd = karatsuba(b, d)
        ad_plus_bc = karatsuba(a + b, c + d) - ac - bd


        return ac*10**(2 * half) + ad_plus_bc*10**(half) + bd

# The main function
a = int(input("Enter first number = "))
b = int(input("Enter second number = "))

# Calling Karatsuba function using these two numbers
print("The product of the two numbers is =",karatsuba(a,b))  